File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/payment/application/query/PaymentQuery.java
package com.example.accountservice.payment.application.query;

import com.example.accountservice.payment.domain.PaymentFindQuery;
import com.example.accountservice.payment.domain.PaymentsWithCount;
import java.time.LocalDateTime;

/**
 * A read-only interface dedicated to the Query Service. A narrow contract separated from the
 * write-side {@code PaymentRepository} (domain). The Query Service should depend only on this
 * interface — it does not expose write methods such as {@code savePayment} (see cqrs-pattern.md).
 */
public interface PaymentQuery {
    PaymentsWithCount findPayments(PaymentFindQuery query);

    /**
     * Card usage statistics (count/total) for a given card — the Card BC's monthly card usage
     * notification (scheduling.md Feature 2) calls this method via {@code PaymentAdapter} (ACL,
     * card/application/adapter/PaymentAdapter). Only COMPLETED payments are aggregated —
     * PENDING/FAILED/CANCELLED were never actually charged, so they are excluded from the
     * statistics. {@code to} is an exclusive upper bound.
     */
    PaymentUsageSummary summarizeCardUsage(String cardId, LocalDateTime from, LocalDateTime to);
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/payment/application/query/PaymentQuery.kt
package com.example.accountservice.payment.application.query

import com.example.accountservice.payment.domain.Payment
import com.example.accountservice.payment.domain.PaymentFindQuery

/**
 * The read-only port — the narrow interface the Query Service depends on (same convention as
 * account/card's `*Query`). Separated from the write model
 * ([com.example.accountservice.payment.domain.PaymentRepository]) to enforce at compile time that the
 * Query Service cannot access write methods like savePayment.
 *
 * `findPayments` reuses `PaymentFindQuery`, as defined by PaymentRepository (the write model), as-is
 * (the same pattern as AccountQuery/CardQuery) — the list query (`GET /payments`) uses only
 * `ownerId`, while a single-record lookup uses the same method with `paymentId` + `ownerId` +
 * `take = 1`.
 */
interface PaymentQuery {
    fun findPayments(query: PaymentFindQuery): Pair<List<Payment>, Long>
}
